### PR TITLE
Miscellaneous improvements to WAL archiving/PITR

### PR DESCRIPTION
**Pull request by soumyadeep2007**
This PR comprises of two commits.

The first one removes some unnecessary diffs (some of which are
historical) in xlog.c by performing a re-merge against the 12 merge
iteration PG parent: 9e1c9f95942.

The second one updates the gpdb_pitr test suite to the post 12 merge
world. There is an [upstream bug](https://www.postgresql.org/message-id/flat/CAE-ML%2B_EjH_fzfq1F3RJ1%3DXaaNG%3D-Jz-i3JqkNhXiLAsM3z-Ew%40mail.gmail.com) that hinders it from fully passing. The
plan is to backport that fix, when it is accepted upstream and then add
this test suite to CI.

More details in the commit messages.

Co-authored-by: Kevin Yeap <kyeap@vmware.com>

### CODE REVIEW

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -8641,7 +8611,7 @@ GetLastSegSwitchData(XLogRecPtr *lastSwitchLSN)
  * This must be called ONCE during postmaster or standalone-backend shutdown
  */
 void
-ShutdownXLOG(int code pg_attribute_unused() , Datum arg pg_attribute_unused() )
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:8644]: we really don't use those arguments, so I won't prefer to loose this even though upstream doesn't have it.

**Review by ashwinstar — COMMENTED** [review]
So, we still can't run the tests. Just getting them up to speed on the upstream, correct?
Once we have the fix we will enable them in CI, right?

**Comment by soumyadeep2007** [comment]
> So, we still can't run the tests. Just getting them up to speed on the upstream, correct?
> Once we have the fix we will enable them in CI, right?

The tests run now. The last test doesn't pass because of the aforementioned bug.
We can sidestep the failure even if we modify the last validate test to
connect in utility mode to look at results instead of setting up
dispatch, but that will break the true spirit of the test - that a GP
cluster can behave as a full GP cluster after a PITR promote.

The tests pass with the patch applied (from the upstream discussion).

So, keeping it out of CI for now and will add it with the backport fix
when its accepted.

**Review by ashwinstar — APPROVED** [review]
Awesome :+1:

**Pull request merged**